scanner/src/main/java/ca/on/oicr/gsi/runscanner/scanner/processor/RevioPacBioProcessor.java

```diff
@@ -0,0 +1,399 @@
+    }
+
+    return dto;
+  }
+
+  /**
+   * Parse a metadata XML file and put all the relevant data into the DTO.
+   *
+   * @param metadata the path to the XML file
+   * @param dto the DTO to update
+   */
+  private void processMetadata(Document metadata, PacBioNotificationDto dto, TimeZone timeZone) {
+    for (RevioPacBioProcessor.ProcessMetadata processor : REVIO_METADATA_PROCESSORS) {
+      try {
+        processor.accept(metadata, dto, timeZone);
+      } catch (XPathException e) {
+        log.error("Failed to extract metadata", e);
+      }
+    }
+  }
+
+  /**
+   * Tests whether a String is blank (empty or just spaces) or null. Duplicated from MISO's
+   * LimsUtils.
+   *
+   * @param s String to test for blank or null
+   * @return true if blank or null String provided
+   */
+  private static boolean isStringBlankOrNull(String s) {
+    return s == null || s.trim().isEmpty();
+  }
+
+  @Override
+  public PathType getPathType() {
+    return PathType.DIRECTORY;
+  }
+
+  /**
+   * Returns a stream that includes Metadata directory level to be used for further processing
+   *
+   * @param runDirectory which we are currently processing
+   * @param directoryName specific sub-directory we want
+   * @return Stream from Metadata directory level
+   */
+  private Stream<File> streamSmrtCellSubdirectories(File runDirectory, String directoryName) {
+    return Stream.of(runDirectory.listFiles())
+        .filter(
+            cellDirectory ->
+                cellDirectory.isDirectory() && REVIO_CELL_DIRECTORY.test(cellDirectory.getName()))
+        .flatMap(
+            cellDirectory ->
+                Stream.of(cellDirectory.listFiles())
+                    .filter(
+                        subDirectory ->
+                            subDirectory.isDirectory()
+                                && subDirectory.getName().equals(directoryName)));
+  }
+
+  /**
+   * Returns the time a file was created
+   *
+   * @param file which we are currently processing
+   * @return Instant point in time
+   */
+  private static Instant getFileCreationTime(File file) {
+    try {
+      FileTime fileTime =
+          (FileTime) Files.getAttribute(Path.of(file.getAbsolutePath()), "creationTime");
+      return fileTime.toInstant();
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * Grab the completion time from a log file
+   *
+   * @param file we are checking
+   * @return completion time
+   */
+  private static Instant getLogCompletionTime(File file) {
+    try {
+      Scanner myReader = new Scanner(file);
+      Pattern pattern =
+          Pattern.compile("^\\[INFO] (.*) \\[.*] exiting with return code \\d+" + " .*");
+      while (myReader.hasNextLine()) {
+        Matcher matcher = pattern.matcher(myReader.nextLine());
+        if (matcher.matches()) {
+          String stringDate = matcher.group(1);
+          String datePattern = "yyyy-MM-dd HH:mm:ss[.SSS][,SSS][Z][X]";
```

Comment:
- All the examples you've seen have had "Z" for the time offset, right? You can put `XX` for the time offset instead of `[Z][X]` if so - I think that's a better assumption than either `X` or `Z`, and I think both definitely shouldn't be there, and it shouldn't be optional unless you've seen a case where it's missing
- Have you seen both "." and "," to separate the milliseconds? If no, only include the one you've seen, and don't make it optional unless you've seen it missing. If you've seen both, I guess just make the separators optional instead of grouping the `SSS` in twice (i.e. `[,][.]SSS`). I don't think you can do an "or" with date formats unfortunately